dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+SPEC(, (, ...), 1)
+SPEC(, (), 0)
+
+}  // namespace detail
+
+template <
+    class GlobRandomIt,
+    class SortableHash =
+        detail::identity_t<typename GlobRandomIt::value_type&>>
+void sort(
+    GlobRandomIt begin,
+    GlobRandomIt end,
+    SortableHash sortable_hash = SortableHash())
+{
+  using iter_type    = GlobRandomIt;
+  using pattern_type = typename iter_type::pattern_type;
+  using value_type   = typename iter_type::value_type;
+  using mapped_type  = typename std::decay<
+      typename detail::closure_traits<SortableHash>::result_type>::type;
+
+  static_assert(
+      std::is_arithmetic<mapped_type>::value,
+      "Only arithmetic types are supported");
+
+  auto pattern = begin.pattern();
+
+  if (pattern.team() == dash::Team::Null()) {
+    DASH_LOG_TRACE("dash::sort", "Sorting on dash::Team::Null()");
+    return;
+  }
+  else if (pattern.team().size() == 1) {
+    DASH_LOG_TRACE("dash::sort", "Sorting on a team with only 1 unit");
+    std::sort(begin.local(), end.local());
+    return;
+  }
+
+  if (begin >= end) {
+    DASH_LOG_TRACE("dash::sort", "empty range");
+    pattern.team().barrier();
+    return;
+  }
+
+  dash::Team& team   = pattern.team();
+  auto const  nunits = team.size();
+  auto const  myid   = team.myid();
+
+  auto const unit_at_begin = pattern.unit_at(begin.pos());
+
+  // local distance
+  auto const l_range = dash::local_index_range(begin, end);
+
+  auto l_mem_begin = begin.globmem().lbegin();
+
+  auto const n_l_elem = l_range.end - l_range.begin;
+
+  auto lbegin = l_mem_begin + l_range.begin;
+  auto lend   = l_mem_begin + l_range.end;
+
+  auto const sort_comp = [&sortable_hash](
+                             const value_type& a, const value_type& b) {
+    return sortable_hash(const_cast<value_type&>(a)) <
+           sortable_hash(const_cast<value_type&>(b));
+  };
+
+  // initial local sort
+  std::sort(lbegin, lend, sort_comp);
+  // Temporary local buffer (sorted);
+  std::vector<value_type> const lcopy(lbegin, lend);
+
+  auto const lmin = (n_l_elem > 0) ? sortable_hash(*lbegin)
+                                   : std::numeric_limits<mapped_type>::max();
+  auto const lmax = (n_l_elem > 0) ? sortable_hash(*(lend - 1))
+                                   : std::numeric_limits<mapped_type>::min();
+
+  dash::Shared<dash::Atomic<mapped_type>> g_min(dash::team_unit_t{0}, team);
+  dash::Shared<dash::Atomic<mapped_type>> g_max(dash::team_unit_t{0}, team);
+
+  g_min.get().op(dash::min<mapped_type>(), lmin);
+  g_max.get().op(dash::max<mapped_type>(), lmax);
+
+  using array_t = dash::Array<std::size_t>;
+
+  std::size_t gsize = nunits * NLT_NLE_BLOCK * 2;
+
+  array_t g_nlt_nle(gsize, dash::BLOCKCYCLIC(NLT_NLE_BLOCK), team);
+
+  array_t g_partition_data(nunits * nunits * 3, dash::BLOCKED, team);
+  std::fill(g_partition_data.lbegin(), g_partition_data.lend(), 0);
+  DASH_ASSERT_GT(
+      g_partition_data.size(),
+      nunits * 2 * 2,
+      "array size is not large enough");
```

Comment:
Why is this constraint needed? Should it be documented?

Reply:
I think it is a relict of old code. Removed now.